components/places/src/db/db.rs

```diff
@@ -131,71 +127,63 @@ impl Deref for PlacesDb {
+        1,
+        true,
+        sql_fns::strip_prefix_and_userinfo,
+    )?;
+    c.create_scalar_function("reverse_host", 1, true, sql_fns::reverse_host)?;
+    c.create_scalar_function("autocomplete_match", 10, true, sql_fns::autocomplete_match)?;
+    c.create_scalar_function("hash", -1, true, sql_fns::hash)?;
+    Ok(())
+}
 
-        host.make_ascii_lowercase();
-        let mut rev_host_bytes = host.into_bytes();
-        rev_host_bytes.reverse();
-        rev_host_bytes.push(b'.');
+mod sql_fns {
+    use crate::api::matcher::{split_after_host_and_port, split_after_prefix};
+    use crate::hash;
+    use crate::match_impl::{AutocompleteMatch, MatchBehavior, SearchBehavior};
+    use rusqlite::{functions::Context, types::ValueRef, Error, Result};
 
-        let rev_host = String::from_utf8(rev_host_bytes)
-            .map_err(|err| rusqlite::Error::UserFunctionError(err.into()))?;
-        Ok(rev_host)
-    })?;
-    c.create_scalar_function("autocomplete_match", 10, true, move |ctx| {
-        // Eventually we'll be able to borrow out of `ctx`, and avoid many of these copies.
-        let search_string = ctx.get::<String>(0)?;
-        let url = ctx.get::<String>(1)?;
-        let title = ctx.get::<Option<String>>(2)?.unwrap_or_default();
-        let tags = ctx.get::<Option<String>>(3)?.unwrap_or_default();
-        let visit_count = ctx.get::<u32>(4)?;
-        let typed = ctx.get::<bool>(5)?;
-        let bookmarked = ctx.get::<bool>(6)?;
-        let open_page_count = ctx.get::<Option<u32>>(7)?.unwrap_or(0);
-        let match_behavior = ctx.get::<MatchBehavior>(8)?;
-        let search_behavior = ctx.get::<SearchBehavior>(9)?;
+    // Helpers for define_functions
+    fn get_raw_str<'a>(ctx: &'a Context, fname: &'static str, idx: usize) -> Result<&'a str> {
+        ctx.get_raw(idx).as_str().map_err(|e| {
+            Error::UserFunctionError(format!("Bad arg {} to '{}': {}", idx, fname, e).into())
+        })
+    }
 
-        let matcher = AutocompleteMatch {
-            search_str: &search_string,
-            url_str: &url,
-            title_str: &title,
-            tags: &tags,
-            visit_count,
-            typed,
-            bookmarked,
-            open_page_count,
-            match_behavior,
-            search_behavior,
-        };
-        Ok(matcher.invoke())
-    })?;
-    c.create_scalar_function("hash", -1, true, move |ctx| {
+    fn get_raw_opt_str<'a>(
+        ctx: &'a Context,
+        fname: &'static str,
+        idx: usize,
+    ) -> Result<Option<&'a str>> {
+        let raw = ctx.get_raw(idx);
+        if raw == ValueRef::Null {
+            return Ok(None);
+        }
+        Ok(Some(raw.as_str().map_err(|e| {
+            Error::UserFunctionError(format!("Bad arg {} to '{}': {}", idx, fname, e).into())
+        })?))
+    }
+
+    // Note: The compiler can't meaningfully inline these, but if we don't put
+    // #[inline(never)] on them they get "inlined" into a temporary Box<FnMut>,
```

Comment:
Ewww... But thanks for adding the comment explaining it!

Reply:
It's not super important tbh, but leaving them off saves us basically nothing, and I ended up adding them every time I went to profile this.

Reply:
Oh, the "ewww" was directed at the compiler, not your code! 😅